plugins/metalink/utils.c

```diff
@@ -397,28 +397,42 @@ TDNFCheckRepoMDFileHashFromMetalink(
         dwError = TDNFGetResourceType(currHashInfo->type, &currHashType);
         BAIL_ON_TDNF_ERROR(dwError);
 
-        if ((hash_Type > currHashType)||
-           (!TDNFCheckHexDigest(currHashInfo->value, hash_ops[currHashType].length)))
-        {
-            continue;
-        }
-        hash_Type = currHashType;
-        hashInfo = currHashInfo;
+        if (hash_Type < currHashType)
+            hash_Type = currHashType;
     }
 
-    if (hashInfo != NULL)
-    {
-        dwError = TDNFChecksumFromHexDigest(hashInfo->value, digest);
-        BAIL_ON_TDNF_ERROR(dwError);
-
-        dwError = TDNFCheckHash(pszFile, digest, hash_Type);
+    if (hash_Type < 0) {
+        /* no hash type was found */
+        dwError = ERROR_TDNF_INVALID_REPO_FILE;
         BAIL_ON_TDNF_ERROR(dwError);
     }
-    else
+    /* otherwise hash_Type is the best one */
+
+    /* now check for all best hash types. Test until one succeeds
+       or until we run out */
+    for(hashList = ml_ctx->hashes; hashList; hashList = hashList->next)
     {
-        dwError = ERROR_TDNF_INVALID_REPO_FILE;
+        int currHashType = TDNF_HASH_SENTINEL;
+        currHashInfo = hashList->data;
+
+        dwError = TDNFGetResourceType(currHashInfo->type, &currHashType);
         BAIL_ON_TDNF_ERROR(dwError);
+
+        /* filter for our best type and also check that the value is valid */
+        if (hash_Type == currHashType &&
+            TDNFCheckHexDigest(currHashInfo->value, hash_ops[currHashType].length)) {
+            dwError = TDNFChecksumFromHexDigest(currHashInfo->value, digest);
```

Comment:
If current highest priority checksum is not valid then we should continue. That might be the case if we have multiple same priority cheksum the next one might be valid one.

Except above LGTM

Reply:
Okay, good point. I will change it.

Reply:
Thinking about this again - I think now the code is correct. The function `TDNFCheckHexDigest()` checks if the hash is syntactically correct, but not if it matches. We skip if not (this only happens if the metalink file is not correct), and go on to the next one (if any). The function `TDNFChecksumFromHexDigest()` retrieves the hash. If it fails, something serious is wrong and we bail out.

The function `TDNFCheckHash()` then finally checks if the checksum matches. In case it does not, it returns `ERROR_TDNF_CHECKSUM_VALIDATION_FAILED`, and we go on to the next one (if any). If it matches, it returns `0` and we break out of the loop - it means we found a match, no need to check any further. If it returns any thing else, something is wrong and we bail.

So, I don't see anything wrong with this logic.